packages/framework/src/Model/Advert/AdvertRepository.php

```diff
@@ -40,26 +41,41 @@ public function findById($advertId)
     /**
      * @param string $positionName
      * @param int $domainId
+     * @param \Shopsys\FrameworkBundle\Model\Category\Category|null $category
      * @return \Doctrine\ORM\QueryBuilder
      */
-    protected function getAdvertByPositionQueryBuilder($positionName, $domainId)
+    protected function getAdvertByPositionQueryBuilder($positionName, $domainId, $category = null)
     {
-        return $this->em->createQueryBuilder()
+        if ($positionName === AdvertPositionRegistry::POSITION_PRODUCT_LIST && $category === null) {
+            DeprecationHelper::trigger('Retrieving advert on product list page without setting category is deprecated and will be disabled in next major.');
+        }
+
+        $queryBuilder = $this->em->createQueryBuilder()
             ->select('a')
             ->from(Advert::class, 'a')
             ->where('a.positionName = :positionName')->setParameter('positionName', $positionName)
             ->andWhere('a.hidden = FALSE')
             ->andWhere('a.domainId = :domainId')->setParameter('domainId', $domainId);
+
+        if ($category !== null) {
+            $queryBuilder
+                ->leftJoin('a.categories', 'c')
+                ->andWhere('c IN(:category) OR c IS NULL')
```

Comment:
1. Space is missing after `IN`. 
2. In case of single category is it good to use `IN` rather than `c = :category`?
3. I would swap the order of those conditions to be sure it's not `NULL` before comparing it. 

```suggestion
                ->andWhere('c IS NULL OR c = :category')
```

Reply:
Good point